updater: detect conda

n3hrox opened the pull request:
* [x] Have you followed the guidelines in our
      [Contributing document](https://dvc.org/doc/user-guide/contributing/core)?

* [ ] Does your PR affect documented changes or does it add new functionality
      that should be documented? If yes, have you created a PR for
      [dvc.org](https://github.com/iterative/dvc.org) documenting it or at
      least opened an issue for it? If so, please add a link to it.

-----
Resolves https://github.com/iterative/dvc/issues/2611

n3hrox commented:
I have no idea how to get DeepSource to pass these minor issues. Recommended usage by python docs is `subprocess.run`. I've even tried using `os.` that is already imported and used as `os.system` but even moving this code to other place causes issue. Any idea how I could omit that? I don't see any way to check for conda without using some kind of system call.

shcheklein commented on dvc/updater.py:
```diff
@@ -164,23 +179,32 @@ def _get_linux(self):
+            # NOTE: both pkg and cask put dvc binary into /usr/local/bin,
+            # so in order to know which method of installation was used,
+            # we need to actually call `brew cask`
+            ret = os.system("brew cask ls dvc")
```
let's create a ticket to update this - we don't use cask anymore, DVC is part of the official brew now

n3hrox replied:
Should I create the ticket? Does it mean that detecting how `dvc` was installed in brew case is related to path checking as well?

shcheklein commented:
@n3hrox thanks for the PR!

It's not the requirement to fix all the deepsource complaints.

@sanketsaurav may be this https://github.com/iterative/dvc/pull/2679#issuecomment-546716270 can be of some interest to you and you can recommend us a better way to deal with this.

shcheklein commented on dvc/updater.py:
```diff
@@ -145,11 +146,25 @@ def _get_update_instructions(self):
+        return output.read().lower()
+
+    @staticmethod
+    def _get_conda(path):
```
better use `_is_conda` and return boolean

n3hrox commented:
Added test and refactored `_get_conda` to `_is_conda`. Let me know if this is sufficient. 
This PR is the first part for https://github.com/iterative/dvc/issues/2672 which I would also love to do if I find some spare time

n3hrox commented on dvc/updater.py:
```diff
@@ -164,23 +179,29 @@ def _get_linux(self):
-        # NOTE: both pkg and cask put dvc binary into /usr/local/bin,
-        # so in order to know which method of installation was used,
-        # we need to actually call `brew cask`
-        ret = os.system("brew cask ls dvc")
```
Removed since `dvc` is part of brew now. Brew installation should be checked now in different manner as @shcheklein said

shcheklein replied:
thanks, we can remove https://github.com/iterative/dvc/blob/82b69a5abd5278af24b0817092a35ea24f936e4b/dvc/updater.py#L126, right? and may be some cask tests. It's not related to this PR, but since you have started doing this :) Up to you.

shcheklein commented on dvc/updater.py:
```diff
@@ -164,23 +179,29 @@ def _get_linux(self):
+        if __file__.startswith("/usr/local/Cellar"):
+            package_manager = "formula"
+
+        dvc_path = self._get_dvc_path("darwin")
```
would it be enough to analyze `__file__` like a couple lines above? instead of running `which`\`where`? Also, is there a more precise heuristic than `'conda' in`? I think it can easily lead to a lot of false positives - `my-conda-project/.env/bin/dvc` (installed via pip in virtualenv - the recommended way)

n3hrox replied:
But that would also mean that I'm checking only the default conda path. Also the paths end differently depending on if it is `conda` or `miniconda`. If you don't mind I try to solve this on Friday because most likely I won't find time during week for that. @shcheklein I hope you are fine with that

shcheklein replied:
@n3hrox absolutely! take your time.

efiop commented on dvc/utils/pkg.py:
```diff
@@ -1,6 +1,15 @@
 
+def is_conda():
+    try:
+        from .build import PKG  # patched during conda package build
```
When we start including this file into other packages, we would be able to simply delete this `dvc/utils/pkg.py` and have `build.py`(or maybe `pkg.py` if we decide to rename it) have `PKG = "something"` defined and then just `from .build import PKG` and use it instead of `get_package_manager`. That would look and feel very nice. :slightly_smiling_face:

n3hrox replied:
@efiop I will rework that method in the next PR where I include patches for other builds as well ;)
